### PR TITLE
Added tests for pkg/scheduler/scheduler.go

### PR DESCRIPTION
**Pull request by anujagrawal699**
**Description:** 
This PR significantly enhances the unit tests for the scheduler package. The improvements aim to increase code reliability, maintainability, and test coverage for various scheduler functions.

**Modifications:**
1. pkg/scheduler/scheduler_test.go

**Key Improvements:**
1. Added  tests for `doSchedule`, `doScheduleBinding`, and `doScheduleClusterBinding` functions.
2. Added tests for `scheduleResourceBindingWithClusterAffinity` and `scheduleResourceBindingWithClusterAffinities`.
3. Added tests for `scheduleClusterResourceBindingWithClusterAffinity` and `scheduleClusterResourceBindingWithClusterAffinities`.
4. Added tests for `patchScheduleResultForResourceBinding`.
5. Added tests for the worker functionality and `scheduleNext` method.
6. Added tests for `recordScheduleResultEventForResourceBinding`.
7. Added tests for the `placementChanged` function.
8. Added new test cases for the `TestCreateScheduler`

**Test Coverage:**
1. pkg/scheduler/scheduler_test.go: 29.71% to 66.90%

**What type of PR is this?**
/kind feature

**Which issue(s) this PR fixes**:
Fixes a part of #5470 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -18,23 +18,1232 @@ package scheduler
...
+				assert.NoError(t, err)
+			}
+
+			actions := fakeClient.Actions()
+			patchActions := filterPatchActions(actions)
+
+			if tt.expectSchedule {
+				assert.Len(t, patchActions, len(tt.expectedPatches), "Expected %d patch actions, got %d", len(tt.expectedPatches), len(patchActions))
+				for i, expectedPatch := range tt.expectedPatches {
+					if i < len(patchActions) {
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler_test.go:389]: Why add this judgment?

In addition, the judgment logic of the TestDoScheduleClusterBinding and TestDoScheduleBinding functions is different. Can the TestDoScheduleClusterBinding and TestDoScheduleBinding functions directly determine the binding field obtained at the same time?

File: pkg/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -18,23 +18,1232 @@ package scheduler
...
+				eventRecorder: fakeRecorder,
+				Algorithm:     mockAlgorithm,
+			}
+
+			err := s.scheduleResourceBindingWithClusterAffinities(tt.binding)
+
+			if (err != nil) != tt.expectError {
+				t.Errorf("scheduleResourceBindingWithClusterAffinities() error = %v, expectError %v", err, tt.expectError)
+			}
+		})
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler_test.go:639]: The result is not judged later. The test method can be the same as that of the TestScheduleResourceBindingWithClusterAffinity.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~, It's a big mr.

The name of the new test method is different from that of the existing test function.

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (anujagrawal699) [reply pkg/scheduler/scheduler_test.go:389]: I'll remove those judgements and refractor both tests to be consistent.

**Comment by anujagrawal699** [comment]
> Thanks~, It's a big mr.
> 
> The name of the new test method is different from that of the existing test function.

Yes it is big and the scheduler package is hard to test.

**Comment by XiShanYongYe-Chang** [comment]
> Yes it is big and the scheduler package is hard to test.

You can consider splitting pr.

**Comment by anujagrawal699** [comment]
> > Yes it is big and the scheduler package is hard to test.
> 
> You can consider splitting pr.

I had a question. As i have used mocks, if i spilt the file than the CI tests might fail. That's the reason i tried to make a single PR for a single file. How can i tackle it?

**Comment by XiShanYongYe-Chang** [comment]
> As i have used mocks, if i spilt the file than the CI tests might fail. 

We need to look at what error caused the ci failure. Generally, CI failures are recorded in clear logs to tell us how to modify them.

**Comment by XiShanYongYe-Chang** [comment]
Hi @anujagrawal699, do we still need this pr?

**Comment by anujagrawal699** [comment]
> Hi @anujagrawal699, do we still need this pr?

No, i'll close it. Thanks.

**Comment by anujagrawal699** [comment]
File covered by #5680 and #5681.

**Pull request closed without merging**